File: packages/sdk/go/fingerprint.go
package bugwatch

import (
	"crypto/sha256"
	"fmt"
)

// GenerateFingerprint creates a SHA256-based fingerprint from the event's
// exception type, value, and the first in-app stack frame's filename and function.
func GenerateFingerprint(event *ErrorEvent) string {
	var exType, exValue, frameFilename, frameFunction string

	if event.Exception != nil {
		exType = event.Exception.Type
		exValue = event.Exception.Value

		for _, frame := range event.Exception.Stacktrace {
			if frame.InApp {
				frameFilename = frame.Filename
				frameFunction = frame.Function
				break
			}
		}
	}

	input := exType + exValue + frameFilename + frameFunction
	hash := sha256.Sum256([]byte(input))
	return fmt.Sprintf("%x", hash)
}

File: packages/sdk/php/src/Fingerprint.php
<?php

declare(strict_types=1);

namespace Bugwatch;

use Bugwatch\Types\ErrorEvent;

final class Fingerprint
{
    public static function generate(ErrorEvent $event): string
    {
        $parts = [];

        if ($event->exception !== null) {
            $parts[] = $event->exception->type;
            $parts[] = $event->exception->value;

            if (count($event->exception->stacktrace) > 0) {
                $topFrame = $event->exception->stacktrace[0];
                $parts[] = $topFrame->filename ?? '';
                $parts[] = $topFrame->function ?? '';
                $parts[] = (string) ($topFrame->lineno ?? 0);
            }
        }

        if ($parts === []) {
            $parts[] = $event->level->value;
            $parts[] = $event->id;
        }

        return hash('sha256', implode('|', $parts));
    }
}
